rust/lance-encoding/src/encoder.rs

```diff
@@ -272,26 +270,37 @@ impl CoreArrayEncodingStrategy {
     }
 }
 
+fn get_dict_encoding_threshold() -> u64 {
+    env::var("LANCE_DICT_ENCODING_THRESHOLD")
+        .ok()
+        .and_then(|val| val.parse().ok())
+        .unwrap_or(100)
+}
+
 // check whether we want to use dictionary encoding or not
 // by applying a threshold on cardinality
-// returns true if cardinality < min(100, num_total_rows)
+// returns true if cardinality < threshold but false if the total number of rows is less than the threshold
 // The choice to use 100 is just a heuristic for now
-fn check_dict_encoding(arrays: &[ArrayRef]) -> bool {
-    let mut unique_values = HashSet::new();
-
+// hyperloglog is used for cardinality estimation
+// error rate = 1.04 / sqrt(2^p), where p is the precision
+// and error rate is 1.04 / sqrt(2^12) = 1.56%
```

Comment:
Does this mean our estimate will be "within 1.56% of true cardinality" or does this mean "1.56% of the time we will get the wrong cardinality"?

In other words, if the true cardinality is 10000 does this mean "our answer is between 9844 and 10156" or does this mean that "we might get an answer like 5000 but only 1.56% of the time"?

Reply:
It means "within 1.56% of true cardinality".  So for a true cardinality 100, we may get `101.5` from its `count` API. Its `count` API actually returns a float as an estimated value, and I cast it to a `u64` (round down), which may make the error rate slightly higher.